src/test/java/org/example/springbootintro/book/BookServiceTest.java

```diff
@@ -0,0 +1,246 @@
+
+        // Then
+        assertEquals(expectedDto, actualDto);
+        verify(bookRepository).save(book);
+    }
+
+    @Test
+    @DisplayName("Find book by existing ID")
+    void findById_ExistingId_ReturnsBookDto() {
+        // Given
+        Long bookId = 1L;
+        Book book = new Book();
+        book.setId(bookId);
+        BookDto expectedDto = new BookDto();
+        expectedDto.setId(bookId);
+
+        // Mocking behavior
+        when(bookRepository.findById(bookId)).thenReturn(Optional.of(book));
+        when(bookMapper.toDto(book)).thenReturn(expectedDto);
+
+        // When
+        BookDto actualDto = bookService.findById(bookId);
+
+        // Then
+        assertEquals(expectedDto, actualDto);
+    }
+
+    @Test
+    @DisplayName("Find book by non-existing ID")
+    void findById_NonExistingId_ThrowsEntityNotFoundException() {
+        // Given
+        Long bookId = 1L;
+
+        // Mocking behavior
+        when(bookRepository.findById(bookId)).thenReturn(Optional.empty());
+
+        // When, Then
+        EntityNotFoundException exception = assertThrows(EntityNotFoundException.class,
+                () -> bookService.findById(bookId));
+        assertEquals("Can't find book by id: " + bookId, exception.getMessage());
+    }
+
+    @Test
+    @DisplayName("Delete book by existing ID")
+    void deleteById_ExistingId_DeletesBook() {
+        // Given
+        Long bookId = 1L;
+        Book book = new Book();
+        book.setId(bookId);
+
+        // Mocking behavior
+        when(bookRepository.existsById(bookId)).thenReturn(true);
+
+        // When
+        bookService.deleteById(bookId);
+
+        // Then
+        verify(bookRepository).deleteById(bookId);
+    }
+
+    @Test
+    @DisplayName("Delete book by non-existing ID")
+    void deleteById_NonExistingId_ThrowsEntityNotFoundException() {
+        // Given
+        Long bookId = 1L;
+
+        // Mocking behavior
+        when(bookRepository.existsById(bookId)).thenReturn(false);
+
+        // When, Then
+        assertThrows(EntityNotFoundException.class, () -> bookService.deleteById(bookId));
+    }
+
+    @Test
+    @DisplayName("Search books by default parameters")
+    void search_ByDefault_ReturnsListOfBookDtos() {
+        // Given
+        String[] titles = {"title1", "title2"};
+        String[] authors = {"author1", "author2"};
+        BookSearchParametersDto params = new BookSearchParametersDto(titles, authors);
+
+        Specification<Book> bookSpecification = mock(Specification.class);
+        when(bookSpecificationBuilder.build(params)).thenReturn(bookSpecification);
+
+        List<Book> books = new ArrayList<>();
+        Pageable pageable = PageRequest.of(0, 10);
+        books.add(new Book());
+        books.add(new Book());
+        Page<Book> page = new PageImpl<>(books);
+        when(bookRepository.findAll(bookSpecification, pageable)).thenReturn(page);
+
+        when(bookMapper.toDto(any())).thenAnswer(invocation -> {
```

Comment:
let's move these `when(...)` mock methods to the `// When` section, same for other tests